### PR TITLE
DirectiveLocations AST node, fix parsing/printing directive locations, set Location

### PR DESCRIPTION
**Pull request by sungam3r**

### CODE REVIEW

File: src/GraphQLParser.Tests/ParserTests.cs [hunk 1]
```diff
@@ -779,6 +779,30 @@ public void Should_Parse_Directives(string text, bool repeatable)
...
+        [InlineData(@"directive @dir on
+OOPS
+| ENUM_VALUE", 2, 1)]
+        [InlineData(@"directive @dir on
+| OOPS
+| ENUM_VALUE", 2, 3)]
+        [InlineData(@"directive @dir on
+|  FIELD_DEFINITION
+|          OOPS", 3, 12)]
+        public void Should_Throw_On_Unknown_DirectiveLocation(string text, int line, int column)
```
Comment (sungam3r) [inline src/GraphQLParser.Tests/ParserTests.cs:798]: This test failed before changes.

File: src/GraphQLParser/AST/Enums/DirectiveLocation.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
+using System.ComponentModel;
+
+namespace GraphQLParser.AST
+{
+    /// <summary>
+    /// Directives must only be used in the locations they are declared to belong in.
+    /// https://graphql.github.io/graphql-spec/June2018/#sec-Type-System.Directives
+    /// </summary>
+    public enum DirectiveLocation
```
Comment (sungam3r) [inline src/GraphQLParser/AST/Enums/DirectiveLocation.cs:9]: Taken from GraphQL.NET so can be removed from there later.

File: src/GraphQLParser/ParserContext.Parse.cs [hunk 3]
```diff
@@ -72,8 +73,13 @@ private GraphQLArgument ParseArgument()
         private GraphQLArguments ParseArguments()
         {
             IncreaseDepth();
+
+            int start = _currentToken.Start;
+
             var args = NodeHelper.CreateGraphQLArguments(_ignoreOptions);
             args.Items = OneOrMore(TokenKind.PAREN_L, (ref ParserContext context) => context.ParseArgument(), TokenKind.PAREN_R);
+            args.Location = GetLocation(start);
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.Parse.cs:81]: I added forgotten location for every new node added in the latest commits.

File: src/GraphQLParser/ParserContext.Parse.cs [hunk 4]
```diff
@@ -326,28 +357,40 @@ private bool ParseRepeatable()
...
+
+            var items = new List<DirectiveLocation>();
 
             // Directive locations may be defined with an optional leading | character
             // to aid formatting when representing a longer list of possible locations
             Skip(TokenKind.PIPE);
 
             do
             {
-                locations.Add(ParseName());
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.Parse.cs:339]: Important thing here - _location_ **is not** a _GraphQLName_. It was a bug. Location is _oneOf_ term - closed set of predefined values. I'll add the same fix for _OperationType_ later in further PR.

File: src/GraphQLParser/ParserContext.Parse.cs [hunk 5]
```diff
@@ -1142,6 +1189,38 @@ private OperationType ParseOperationType()
...
+            {
+                // http://spec.graphql.org/June2018/#ExecutableDirectiveLocation
+                "QUERY" => DirectiveLocation.Query,
+                "MUTATION" => DirectiveLocation.Mutation,
+                "SUBSCRIPTION" => DirectiveLocation.Subscription,
+                "FIELD" => DirectiveLocation.Field,
+                "FRAGMENT_DEFINITION" => DirectiveLocation.FragmentDefinition,
+                "FRAGMENT_SPREAD" => DirectiveLocation.FragmentSpread,
+                "INLINE_FRAGMENT" => DirectiveLocation.InlineFragment,
+                "VARIABLE_DEFINITION" => DirectiveLocation.VariableDefinition,
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.Parse.cs:1205]: New location from October spec.

File: src/GraphQLParser/ParserContext.cs [hunk 6]
```diff
@@ -144,7 +169,36 @@ private void ExpectKeyword(string keyword)
 
         private void Throw_From_ExpectKeyword(string keyword)
         {
-            throw new GraphQLSyntaxErrorException($"Expected \"{keyword}\", found Name \"{_currentToken.Value}\"", _source, _currentToken.Start);
+            throw new GraphQLSyntaxErrorException($"Expected \"{keyword}\", found {_currentToken}", _source, _currentToken.Start);
+        }
+
+        private string ExpectOneOf(string[] oneOf)
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.cs:175]: Can be used for all _oneOf_ terms from the spec.

File: src/GraphQLParser/Visitors/SDLWriter.cs [hunk 7]
```diff
@@ -298,9 +299,24 @@ public override async ValueTask VisitDirectiveDefinition(GraphQLDirectiveDefinit
         {
             await Visit(directiveDefinition.Comment, context).ConfigureAwait(false);
             await Visit(directiveDefinition.Description, context).ConfigureAwait(false);
-            await context.Write("directive ").ConfigureAwait(false);
+            await context.Write("directive @").ConfigureAwait(false);
```
Comment (sungam3r) [inline src/GraphQLParser/Visitors/SDLWriter.cs:302]: oops

File: src/GraphQLParser/Visitors/SDLWriter.cs [hunk 8]
```diff
@@ -681,6 +697,7 @@ public override async ValueTask VisitArgument(GraphQLArgument argument, TContext
         public override async ValueTask VisitArgumentsDefinition(GraphQLArgumentsDefinition argumentsDefinition, TContext context)
         {
             await context.Write("(").ConfigureAwait(false);
+            await context.WriteLine().ConfigureAwait(false);
```
Comment (Shane32) [inline src/GraphQLParser/Visitors/SDLWriter.cs:700]: ?? This isn't typical.  At least, most graphql formatting tools, like 'prettify' in GraphiQL, would not do this

File: src/GraphQLParser/Visitors/SDLWriter.cs [hunk 9]
```diff
@@ -854,7 +902,7 @@ private int GetLevel(TContext context)
                         ++level;
                 }
 
-                if (currentNode is GraphQLDescription)
+                if (currentNode is GraphQLDescription || currentNode is GraphQLDirectiveDefinition)
```
Comment (sungam3r) [inline src/GraphQLParser/Visitors/SDLWriter.cs:905]: I still do not know how to do it better with indentation (and new lines). I'll think about that later after all PRs around grammar.

File: src/GraphQLParser/Visitors/SDLWriter.cs (hunk 8 above)
Comment (sungam3r) [reply src/GraphQLParser/Visitors/SDLWriter.cs:700]: https://github.com/graphql-dotnet/parser/pull/160/files/b1f15e0490ece23f5d820497450d809509c11857#r778103802

**Review by Shane32 — APPROVED** [review]

**Pull request merged**